text/0000-target-bundles.md

```diff
@@ -0,0 +1,245 @@
+- Feature Name: target_bundles
+- Start Date: 2016-08
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Combine distribution of standard libraries and targets into bundles for targeting a particular
+platform. Such bundle is all you need from the Rust side to cross-compile for your target.
+
+The major ideas of this RFC is to:
+
+1. Make JSON targets as full featured as they need to be in order to support specifying targets as
+   custom as they need to be;
+2. Convert current built-in targets to the JSON targets;
+3. Change distribution of libstd to include a corresponding JSON target.
+
+# Motivation
+[motivation]: #motivation
+
+Currently there’s two different ways rustc targets are distributed: built-in targets and custom
+JSON targets. Built-in targets are very inflexible – they cannot be changed without changing and
+recompiling the compiler itself. On the other side of the spectrum, custom JSON targets are easy to
+adjust and adapt, but feature wise are very limited and rarely are suitable for the more uncommon
+use-cases.
+
+We’ve observed a considerable amount of desire by users of the language to customize targets they
```

Comment:
I don't think it makes sense to suffix the rustlib-target-dir with a hash. Consider that if we have a directory blessed to contain target specifications, and the name of those specifications is `<target-name>.ext`, we already know that all the _installed_ specfications have unique names.

If we're worried about handling targets that don't have specifications installed, why would they have `std` installed in `rustlib`? Instead, for handling non-installed target specs (those not in the blessed dir), I'd propose the following:

 - allow specification of the directory containing `std` crates either in the target-spec (brittle for dynamic uses, but potentially automatic) or as a seperate argument to rustc (easy for dynamic, but no automation)

Alternately (and I prefer this):

 - allow `--target` to accept a directory (so `--target ./my-target`) that contains a target spec (`./my-target/target.json` or `./my-target/target.toml`) and the `std` crates in the same way as `rustlib` today.